### PR TITLE
[Blog app] Milestone 11 - Authorization with CanCanCan

### PR DESCRIPTION
**Pull request by cherelemma**
##  Hello Dear CR ✋ 

I hope this message finds you well. I am submitting a pull request for your review and feedback on the implementation of the **Blog app - Authorization with CanCanCan**. I, @cherelemma, and @misalazeem collaborate to implement the project requirements.  We believe the implementation fulfills the project requirements and also maintains the general requirements.

## Project Requirements ✔️

- Install CanCanCan in your project.
- Add a `role` column to the users table. Remember to use a migration for this.
- A user can delete a post if it is theirs or if they have an admin role (column `role` has value `"admin"`). Use CanCanCan for this authorization.
    - For that you need to implement the post deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.
- A user can delete a comment if it is theirs or if they have an admin role (column `role` has value `"admin"`). Use CanCanCan for this authorization.
    - For that you need to implement the comment deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.

## Implementation Summary ✔️

In this milestone, we completed the following tasks to implement the project requirements.

➡️Install CanCanCan gem
➡️Generate the CanCanCan mode 
➡️Add a role to the User model
➡️ Create the Ability model and set up roles for the users
➡️Add load and authorization to controllers
➡️Create conditional rendering
➡️Add the 'Delete' button for Posts and Comments

## General requirements ✔️

- There are [No linter errors](https://github.com/microverseinc/linters-config)
- Follow correct [GitFlow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md)
- Follow the list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).

We have thoroughly reviewed the project requirements and cross-checked the implementation of user authorization of the Blog app against the given description to ensure accuracy. All necessary views have been run successfully.

We'd like to ask for a review, as your feedback and suggestions are invaluable in ensuring the quality and efficiency of the project. 

Thank you in advance for your time and assistance. We look forward to your feedback and addressing any concerns to enhance the overall quality of the project.

### CODE REVIEW

File: db/migrate/20230814094947_create_users.rb [hunk 1]
```diff
@@ -8,5 +8,6 @@ def change
 
       t.timestamps
     end
-  end
+  
+    add_index :users, :name, unique: true
 end
```
Comment (AmaduKamara) [inline db/migrate/20230814094947_create_users.rb]: - [ ] There is a missing `end` for the class that is causing the migration to not run successfully. Kindly add the closing `end` for the class so your migration for create_users works fine. 👍🏽

File: app/views/partials/_comments.html.erb [hunk 2]
```diff
@@ -8,6 +8,12 @@
       <div class="comment">
           <span class="comment-author"><%= comment.author.name %>:</span>
           <span><%= comment.text %></span>
+
+          <% if can? :destroy, comment %>
+              <%= link_to user_post_comment_url(author_id: comment.post.author.id, post_id: comment.post.id, id: comment.id), method: :delete, data: { confirm: "Are you sure?" } do %>
+                  <button class="short-button">Delete</button>
+              <% end %>
+          <% end %>
```
Comment (AmaduKamara) [inline app/views/partials/_comments.html.erb]: - [ ] I noticed the delete operation for comments is not functioning as it should. Kindly work on it so a comment can be deleted as expected 👍🏽

File: app/views/partials/_counter.html.erb [hunk 3]
```diff
@@ -1,23 +1,31 @@
...
+        <% end %>
+    <% end %>
+
+    <% if can? :destroy, post %>
+        <div class="btn-delete">
+            <%= link_to user_post_url(post.author_id, post), method: :delete, class: 'btn-link' do %>
+                <button>Delete Post</button>
+            <% end %>
+        </div>
+    <% end %>
```
Comment (AmaduKamara) [inline app/views/partials/_counter.html.erb]: - [ ] I noticed the delete operation for a post is not functioning as it should. Kindly work on it so a post can be deleted as expected 👍🏽

**Review by AmaduKamara — CHANGES_REQUESTED** [review]
# Status: Required Changes ♻️ ♻️ ♻️

Hi Team,

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

✅ You used Gitflow correctly 👍🏽
✅ The Linter checks passed successfully 👍🏽
✅ Used CanCan gem 👍🏽
✅ Good commit messages 👍🏽

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

File: db/migrate/20230814094947_create_users.rb (hunk 1 above)
Comment (Abenezer-Tilahun) [reply db/migrate/20230814094947_create_users.rb]: Fixed.

File: app/views/partials/_comments.html.erb (hunk 2 above)
Comment (Abenezer-Tilahun) [reply app/views/partials/_comments.html.erb]: Fixed.

File: app/views/partials/_counter.html.erb (hunk 3 above)
Comment (Abenezer-Tilahun) [reply app/views/partials/_counter.html.erb]: Fixed.

**Review by Abenezer-Tilahun — APPROVED** [review]
Hi @cherelemma @misalazeem 
 
Your project is complete  👍 💯   There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 👏🏻

✔️ No linter errors.
✔️ You used correct Git-Flow.
✔️ Professional README  
✔️ Good Commit Messages
✔️ Good job for Implemented Suggested changes.

## Optional suggestions 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

**Pull request merged**